### DISCUSSION TITLE
Exclusive constraint using STABLE function

### DISCUSSION THREAD

**Discussion by adriangb** [geldata/gel#2096] [Q&A] [answered]
Hi, I'm trying to do the following:

```
abstract link timed_link {
    property deleted_at -> datetime;
}
type Container {
    multi link links extending timed_link -> Container{
        constraint exclusive on ((__subject__@source, __subject__@target, __subject__@deleted_at > datetime_of_statement()));
    };
}
```

Where I am trying to express that `Container` links to multiple other `Container`, and multiple links can exist between a pair of `Container`s, but only one of those links can be "valid" at a time.

This fails with:

```
db.errors.InternalServerError: functions in index expression must be marked IMMUTABLE
```

This seems to be because `datetime_of_...` are `STABLE` (guessing) functions, not `IMMUTABLE`. That makes sense, I can definitely see how this would create problems since `exclusive` constraints create an index behind the scenes. So is there another way to express this exclusivity constraint if I don't need an index?

**Comment by elprans** [comment, marked as answer]
> and multiple links can exist between a pair of `Container`s, but only one of those links can be "valid" at a time.

You cannot have multiple links of the same name between two specific objects.  There is an implicit `(@source, @target)` constraint which cannot be disabled.  You need an intermediate object for your case, perhaps like this:

```
type ContainerLink {
    required link source_container -> Container;
    required link target_container -> Container;
    property deleted -> bool;
    property deleted_at -> datetime;
}
```

As for the constraint, it is impossible to have an `exclusive` constraint without creating a unique index, because that is how unique constraints are implemented in Postgres.  Here you need a partial constraint:

```
type ContainerLink {
    ...
    constraint exclusive on ((.source_container, .target_container)) except (.deleted)
}
```

Unfortunately, support for partial constraints is not in the current release, but we plan to implement it for the next one. (see #1610)

**Reply by adriangb** [reply]
Thank you for the prompt reply!

> You cannot have multiple links of the same name between two specific objects

Got it, this makes sense.

> You need an intermediate object for your case, perhaps like this

Something like that should work. Inserts become a bit ugly, but I think this can be ameliorated by using some functions like:

```
function link(parent: Container, child: Container) -> Container
    using EdgeQL $$
        UPDATE Container
        FILTER .id = parent.id
        SET {
            children += (
                INSERT ContainerLink {
                    source_container := (
                        SELECT DETACHED Container
                        FILTER .id = parent.id
                    ),
                    target_container := (
                        SELECT DETACHED Container
                        FILTER .id = child.id
                    ),
                }
            )
        }
    $$;
```

Where the rest of the schema is:

```
type ContainerLink {
    required link source_container -> Container;
    required link target_container -> Container;
}
type Container {
    multi link children -> ContainerLink;
}
```

Q: can I call a function from within another? In this case, it would make sense to have a function `unlink(child, parent)` that gets called by `link` to ensure that only 1 valid link exists at a time.

> support for partial constraints is not in the current release, but we plan to implement it for the next one

Great to hear you are planning on that already!

> property deleted -> bool

It seems that in addition to the `deleted_at -> datetime` property I would also need a boolean `deleted` property that needs to be updated at the same time. In other words, some data duplication and client-side logic is needed here because, even when partial constraints are implemented, we will never be able to use a non-IMMUTABLE function like `datetime_of_statement` as part of a constraint?

**Reply by adriangb** [reply]
A small followup @elprans. Your suggestion was great, but it lead to another issue. I don't think that:
```
constraint exclusive on ((.source_container, .target_container)) except (.deleted)
```
Would work for my use case (sorry if I didn't explain it clearly!), since that disallows having multiple _deleted_ `ContainerLink` (i.e. only 1 deleted and 1 non-deleted would be allowed). The original intention was that there should only ever be 1 _non-deleted_ link, but there can be many _deleted_ link.

Small edge case, but I wanted to report it just in case it's helpful.

Thank you for such a great project and for being so active in the discussions!
